llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/Passes.h

```diff
@@ -0,0 +1,27 @@
+//===-- Passes.h - LinalgTransform passes -----------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```

Comment:
Where is this code coming from? If it is LLVM licensed then it should go in IREE's third_party/ directory. Otherwise it should have IREE's license header. I also see some files like https://github.com/google/iree/blob/main/llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/Passes/PassDetail.h without license headers and https://github.com/google/iree/blob/main/llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/Transforms/Utils.h with the LLVM header. 

Those were added in https://github.com/google/iree/pull/8554, which @MaheshRavishankar reviewed.

Reply:
It is coming from https://github.com/google/iree-llvm-sandbox/, added a more descriptive commit message.

I don't have an opinion on the header/license: most of the code in iree-dialects/LinalgExt and LinalgTransform will go to the LLVM repo in the fullness of time (probably O(months)).

Is there a strong opinion that we should impose the IREE header now?
If so, an automated way would be nice..

Reply:
> We should not have any code in IREE that isn't under third_party and is licensed under another license. I see that this has already [happened](https://cs.github.com/google/iree?q=%22part+of+the+llvm+project%22) as part of #8554. Since contributions to that repo are under the Google CLA, we can relicense it here. Besides any other reason to do so, the header saying it is "part of the LLVM project" is not accurate.
> 

It is not licensed under a different license: The header is different. This comes from our project in another repo, covered by the CLA. There is not an issue here aside from a syntactic one.

> For automated processing, you may find https://cs.github.com/google/iree/blob/main/scripts/add_license_header.py useful, although probably a simple multi-file find-replace would be more effective, as there is already a license header present that also needs to be removed.
> 
> This issue actually highlights a general concern I have with this PR and with #8554. We are pulling in large chunks of code from another repository with minimal review. From the README of the sandbox repository:

It has been reviewed by multiple people. Just in our other repo.

> 
> > This repository contains experimental work by the IREE team closely related to LLVM and MLIR, usually with the aim of upstreaming in some form. The main project is at https://github.com/google/iree.
> > As an experimental project, build greenness, documentation, and polish are likely to be minimal, as it instead prioritizes easy experimentation.
> 
> Those disclaimers do not apply to the IREE project. We should not in general be prioritizing easy experimentation at the cost of build greenness, documentation, and polish. That means that code coming from the sandbox should be reviewed just as any new contribution (ideally in small, reviewable chunks) and should meet the general standards and style of the IREE project. For instance (just picking random examples that are easy to spot) it should not be making [heavy use of forward declarations](https://github.com/google/iree/pull/8554/files#diff-a35206acdf5ae002aae66ddfb8def829b1b57a44e36cce923c1f494e943104c7R17) or [using `llvm_unreachable` in in non performance-critical settings](https://github.com/google/iree/pull/8554/files#diff-795d84c60cd560c80c90aecf4ce7ea7fd66d4b941917371b41cb2e690f9112f5R171).

The code in llvm-external-projects is staged for contribution to upstream and we follow the standards of upstream for it without comment.

> 
> I think it's likely there have been some more detailed discussions on this integration strategy that I haven't been a party to, so apologies if I'm retreading ground here, but I think we should take a step back on our strategy for pulling in the advances from the sandbox into the IREE project.

Yes. This is fine and the right way to go.

Reply:
> It is not licensed under a different license: The header is different. This comes from our project in another repo, covered by the CLA. There is not an issue here aside from a syntactic one.

Right, it's good that this is covered under our CLA, which allows us to change the license header. The current header gives the appearance, however, that this comes from the LLVM upstream, which would mean that we *couldn't* change the header. I agree with you that there should only be a syntactic problem here, but it's a syntactic problem that can easily create confusion or result in accidentally creating a real problem.

> The code in llvm-external-projects is staged for contribution to upstream and we follow the standards of upstream for it without comment.

Ok, sure, that makes sense: I withdraw my comment on style (with a final point that regardless of style, an `llvm_unreachable` followed by another statement is just nonsensical, since the other statement is, by definition, unreachable). I picked those things because they're objective and easy to search for.

> It has been reviewed by multiple people. Just in our other repo.

Right, but that other repo has an explicitly different standard of review based on different goals. To be clear, having different standards of review for different code is fine: fast experimentation is valuable and that's why we have a separate repo.

If llvm-external-projects is experimental (as suggested by Mahesh), then I think that needs to be explicitly documented somewhere. It also seems directly at odds with my understanding of llvm-external-project/iree-dialects to be a public API for IREE, which, if anything, should be even more carefully reviewed than project internals.

If this code has already been reviewed by some other mechanism, different from the general reviewing standards in iree-llvm-sandbox documented in that repositories README, then I think that information should be included in the PR description.

Ultimately, this is not my decision. I am merely lodging my concern, which I expect to be overruled.

Reply:
> > It is not licensed under a different license: The header is different. This comes from our project in another repo, covered by the CLA. There is not an issue here aside from a syntactic one.
> 
> Right, it's good that this is covered under our CLA, which allows us to change the license header. The current header gives the appearance, however, that this comes from the LLVM upstream, which would mean that we _couldn't_ change the header. I agree with you that there should only be a syntactic problem here, but it's a syntactic problem that can easily create confusion or result in accidentally creating a real problem.

Agreed. Started an internal thread where I mentioned that the sandbox should have been covered by the same guidance wrt license header as the main project and should probably be brought up to date.

> 
> > The code in llvm-external-projects is staged for contribution to upstream and we follow the standards of upstream for it without comment.
> 
> Ok, sure, that makes sense: I withdraw my comment on style (with a final point that regardless of style, an `llvm_unreachable` followed by another statement is just nonsensical, since the other statement is, by definition, unreachable). I picked those things because they're objective and easy to search for.
> 
> > It has been reviewed by multiple people. Just in our other repo.
> 
> Right, but that other repo has an explicitly different standard of review based on different goals. To be clear, having different standards of review for different code is fine: fast experimentation is valuable and that's why we have a separate repo.
> 
> If llvm-external-projects is experimental (as suggested by Mahesh), then I think that needs to be explicitly documented somewhere. It also seems directly at odds with my understanding of llvm-external-project/iree-dialects to be a public API for IREE, which, if anything, should be even more carefully reviewed than project internals.

The iree_input dialect is such an API. The rest is not.

> 
> If this code has already been reviewed by some other mechanism, different from the general reviewing standards in iree-llvm-sandbox documented in that repositories README, then I think that information should be included in the PR description.
> 
> Ultimately, this is not my decision. I am merely lodging my concern, which I expect to be overruled.

I wouldn't overrule it so much as say that Mahesh is the code owner on this stuff and is well aware of which parts of the joint development in sandbox were experimental and what was developed to the higher standard. I trust his judgment and the context he has on it.

Reply:
> If llvm-external-projects is experimental (as suggested by Mahesh), then I think that needs to be explicitly documented somewhere. It also seems directly at odds with my understanding of llvm-external-project/iree-dialects to be a public API for IREE, which, if anything, should be even more carefully reviewed than project internals.

To clarify. Not all of `llvm-external-projects` is experimental (sorry if I implied that). The things landing within this (`LinalgExt/Transforms`, etc.) are experimental from IREEs stand point. It is not part of IREE public API, and its use in IREE is exploratory at this point, i.e., its use is guarded by flag. This would move to upstream MLIR. In terms of testing, it might not have the standards of testing that `iree/compiler` has, but it is experimental, so I would think that is expected. This will move upstream when it is ready and at that time I would expect it to have the same quality/standards of testing as MLIR. In terms of burden on IREE
- Not part of pulbic API
- If these specific tests break (during integrate, cause core IREE members are not really working on it actively, and they wont break tests anyway) then my expectation is that they get disabled. We can file an issue, but that is also not necessary IMO, more needs an FYI to Nicolas (or anyone working there).

Reply:
That certainly alleviates some of my concerns. Could you (or someone) please add a README to https://github.com/google/iree/tree/main/integrations/tensorflow/iree-dialects/lib/Dialect/LinalgExt documenting this?

Reply:
@nicolasvasilache could you please fix all the license headers in iree-llvm-sandbox, the files from this already checked in to iree: https://cs.github.com/google/iree?q=%22part+of+the+llvm+project%22. That add_license_header.py script should catch any missing the license header and a find replace can fix any that currently have the LLVM header

Reply:
https://github.com/google/iree/pull/8603